text/0026-serialization.md

```diff
@@ -0,0 +1,598 @@
+- FastBufferReader and FastBufferWriter are intended to make data easier to debug - one such thing to support will be a `#define MLAPI_FAST_BUFFER_UNPACK_ALL` that will disable all packing operations to make the buffers for messages that use them easier to read.
+
+## INetworkSerializable
+
+A new `BufferSerializer` class wrapping `FastBufferReader` and `FastBufferWriter` will replace the existing `NetworkSerializer` to provide the same two-way interface as the existing `NetworkSerializer`, and is rewritten as a **ref struct** so that no allocation overhead is incurred in calling the `INetworkSerializable`'s `Serialize()` method. A **ref struct** is chosen instead of a normal struct to avoid the pitfalls associated with having mutable structs and forgetting to pass them with the ref keyword - we can manage this with `FastBufferReader` and `FastBufferWriter` because those are internal and we can rely on our ability to understand the implications of using them as structs, but our end users may be less experienced and may inadvertently pass this forward to calls that do not take the value by ref. A `ref struct` is always passed by reference rather than copied, but does have the disadvantage of being limited to the stack frame - so we can't use it for `FastBufferWriter` and `FastBufferReader` because we persist those.
+
+`BufferSerializer` has an additional advantage (or disadvantage, depending on your viewpoint) compared to `FastBufferWriter` and `FastBufferReader`: since it's intended to be used by our end users, it performs bounds checking on every operation and throws exceptions if attempting to read or write out of bounds. This, combined with the `IsReader` if-checks, will make `BufferSerializer` slower than `FastBufferReader` and `FastBufferWriter` - but for advanced users, we can expose `FastBufferReader` and `FastBufferWriter` and allow the user to use them directly if they choose (with all appropriate warnings about them delivered).
+
+## Custom Messages
+
+Custom messages as they are, using buffers, no longer need to exist. Since the message handling code is part of the message definition, users have the freedom to define their own messages. Our internal code doesn't need to know anything about the messages themselves in order to send or receive them, so users can create any custom message they want the same way we define our own internal messages, and send them the same way we send ours, and their message handler will be magically invoked when the message is received. The only gotcha is the MessageType enum - since users can't extend the enum itself, we will provide a MessageType.Custom that's always at the end of the enum, and users can operate as such:
+
+```C#
+enum CustomMessageType
+{
+	MyMessage1,
+	MyMessage2,
+	Etc,
+}
+
+struct Message : IMessage
+{
+	MessageType GetMessageType()
+	{
+		return MessageType.Custom + (int)CustomMessageType.MyMessage1;
+	}
+
+    int GetSizeUpperBound()
```

Comment:
I’m also wondering if some of the boilerplate could be abstracted away, like GetSizeUpperBound() and GetMessageType() . I get that you can’t have a base class with structs. I’m wondering if that could be calculated for you instead.
As a user writing custom messages, I wouldn't really care about these, I'd just care about my message's content and how I handle it on the receiving end.

Reply:
Not sure how to abstract either of those away.

`GetMessageType()` should probably be a property and not a method, but each and every message struct has to have a different return value from that because that's what goes in the message header and how the handler is looked up on the receiving end.

`GetSizeUpperBound()`... there's no generic implementation for this that would work. The size has to be known before serializing to a buffer, but imagine a message like this:

```C#
struct Message : IMessage
{
    bool HasValue;
    SomeBigStruct Value;
    List<int> VariadicData;
    
    void Serialize(ref FastBufferWriter writer)
    {
        writer.WriteValue(HasValue);
        if(HasValue)
        {
            writer.WriteValue(Value);
        }
        writer.WriteValue(VariadicData.Count);
        foreach(int dataItem in VariadicData)
        {
            writer.WriteValue(dataItem);
        }
   }
}
```

The size of that message is `sizeof(bool) + (HasValue ? sizeof(SomeBigStruct) : 0) + sizeof(int) + sizeof(int) * VariadicData.Count`. I don't see any way that could be guessed in a generic way.

We could avoid the need for `CalculateSizeUpperBound()` by serializing to a temporary buffer and copying after all the data's been written, but that's not really ideal performance-wise...

Reply:
For GetMesssageType, you could use reflection and use the message class itself as an identifier. If I recall correctly, you can get either a string for the full class name or a hash for that Type. Then order all your Types by hash and increase your counter to generate a shorter value to send. Since you'll have the same message types both sides, that number should be the same on both sides?
For GetSize, we need the size before we serialize? Else couldn't we get that info from the result of Serialize?

Reply:
I personally don't want to go down the road of using reflection / type hashes. There's too many ways that kind of magic can go wrong in my experience. When we start doing type hashes, there's too many ways a code change could break that and now the server can't ever support older clients. I think it's better to be explicit about what value you want in your header.

GetSize... the reason for that was message batching. I wanted to know the size before serializing so I could know if it'll fit into the current message batch buffer or if I needed to make a new one. I wanted to do that without incurring the cost of an extra memcpy by serializing to a temporary buffer and then copying it... but ultimately when I updated the IMessage RFC I decided that simplifying the interface was probably more valuable than avoiding that extra copy, so I removed `CalculateSizeUpperBound()`

Can we continue this discussion on that RFC if you have more feedback?

#29

Reply:
(Also please re-read it, I made some other changes as well.)

Reply:
_(FWIW, I also do not like type-based maps/hashes too)_